### PR TITLE
Make Partition family of APIs use reduce_then_scan algorithm 

### PR DESCRIPTION
**Pull request by danhoeflinger**
This PR changes partition family of scan-like algorithms to use reduce_then_scan where it is beneficial.

Moves all algorithm selection decisions to __parallel_partition_copy.

---

This PR is targeted to #1763, to allow for a clean diff, and is a part of the following sequence of PRs meant to be merged in order:

#1769 [MERGED] ~~Relocate __lazy_ctor_storage to utils header~~
#1770 [MERGED] ~~Use __result_and_scratch_storage within scan kernels~~
#1762 Add reduce_then_scan algorithm for transform scan family
#1763 Make Copy_if family of APIs use reduce_then_scan algorithm
#1764 Make Partition family of APIs use reduce_then_scan algorithm (This PR)
#1765 Make Unique family of APIs use reduce_then_scan algorithm

This work is a collaboration between @mmichel11 @adamfidel and @danhoeflinger, and based upon an original prototype by Ted Painter.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (adamfidel) [inline include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:872]: Maybe we also do not need the default for `_Assign` here.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:872]: Fixed.  In truth, we probably don't need to support an arbitrary assignment operator here, but this makes it more consistent and extension easier.

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: fixed

File: include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h
Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h]: fixed.

**Review by julianmi — APPROVED** [review]
LGTM. I would hold off merging this until the main patch was merged.

**Review by danhoeflinger — COMMENTED** [review]
This one also LGTM, after rebasing.

**Review by julianmi — DISMISSED** [review]
LGTM after a rebase.

**Review by mmichel11 — APPROVED** [review]
LGTM

**Pull request merged**
